File: files/100753f.cpp
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <stdio.h>
#include <vector>
#include <math.h>
typedef unsigned long long ULL;

using namespace std;
ULL mulmod(ULL a, ULL b, ULL c){
    ULL x = 0,y = a%c;
    
    while(b>0){
        if(b&1) x = (x+y)%c;
        y = (y<<1)%c;
        b >>= 1;
    }
    
    return x;
}

ULL pow(ULL a, ULL b, ULL c){
    ULL x = 1, y = a;
    
    while(b>0){
        if(b&1) x = mulmod(x,y,c);
        y = mulmod(y,y,c);
        b >>= 1;
    }
    
    return x;
}

bool miller_rabin(ULL p, int it){
    if(p<2) return false;
    if(p==2) return true;
    if((p&1)==0) return false;
    
    ULL s = p-1;
    while(s%2==0) s >>= 1;
    
    while(it--){
        ULL a = rand()%(p-1)+1,temp = s;
        ULL mod = pow(a,temp,p);
        
        if(mod==-1 || mod==1) continue;
        
        while(temp!=p-1 && mod!=p-1){
            mod = mulmod(mod,mod,p);
            temp <<= 1;
        }
        
        if(mod!=p-1) return false;
    }
    
    return true;
}

ULL primes[1000080]={0};
int main()
{
    ULL n=1000000,i,j,k,l,solution=1,coun;

    for(i=2;i<=1000;i++)
        if(!primes[i]){
            for(j=i*i;j<=1000009;j+=i)
                primes[j]++;
    }
    vector<ULL> v;
    for(i=2;i<=1000009;i++)
        if(!primes[i])v.push_back(i);
    cin>>n;
    i=0;
    while(i<v.size())
    {
        if(v[i]*v[i]*v[i]>n)
            break;
        if(n%v[i]==0)
        {
            coun=1;
            while(n%v[i]==0)
                {n=n/v[i];coun++;}
            solution*=coun;
        }
        i++;
    }
    if(n==1)
        solution=1;
    else if(miller_rabin(n,55))solution*=2;
    else if(sqrt(n)*sqrt(n)==n&&miller_rabin(sqrt(n),55))
        solution*=3;
    else solution*=4;
    cout<<solution<<endl;



}

File: files/100753f.rb
def modulo(a,b,c)
        x = 1
        y = a
        while b>0
                if b%2==1
                        x = (x*y)%c
                end        
                y = (y*y)%c
                b = b/2
        end
        return x%c
end 
def millerrabin(n,iteration)
        if n<2
                return 0
        end        
        if n!=2 && n%2==0
                return 0
        end        
         
        d=n-1
        while d%2==0
                d = d/2
        end
        i=1
        while  i<iteration
                a = 1+rand(n-1)
                temp = d
                x = modulo(a,temp,n)
                while temp!=n-1 && x!=1 && x!=n-1
                        x = (x*x)%n
                        temp = temp*2
                end
                if x!=n-1 && temp%2==0
                        return 0
                end        
        i+=1
        end
        
        return 1
end        

primes=Array.new(1000009,0)
fprimes=Array.new(1000009,0)

fcoun=0
i=2
while i<=1001
      if primes[i]==0
        j=i*i
        while j<=1000002
        primes[j]+=1
        j+=i
        end
      end
      i+=1
end 
i=2
while i<=1000002
    if primes[i]==0
        fprimes[fcoun]=i
        fcoun+=1
    end
    i+=1
end

i=0
solution=1
n=gets().to_i
while i<=fcoun
     if fprimes[i]*fprimes[i]*fprimes[i]>n
        break
     end
     if n%fprimes[i]==0
        coun=1
        while n%fprimes[i]==0
              n/=fprimes[i]
              coun+=1
        end
        solution*=coun
     end
     i+=1
end
root=Math.sqrt(n)
if millerrabin(n,20)==1
        puts solution*2
elsif root*root==n&& millerrabin(root,20)==1    
        puts solution*3
else
        puts solution*4
end     
        
        
        
        


